File: 046_Permutations.cpp
//43.5%
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
		if(nums.empty()) return res;
		int size(nums.size());
		vector<int> vec(size, 0);
		permuteHelper(nums, 0, size, res, vec);
		return res;
	}
private:
	void permuteHelper(vector<int>& nums, int left, int right, vector<vector<int>>& res, vector<int>& vec){
		if(left == right){
			res.push_back(vec);
			return;
		}
			
		for(int i = left; i < right; i++){
			swap(nums[i], nums[left]);
			vec[left] = nums[left];
			permuteHelper(nums, left + 1, right, res, vec);
			swap(nums[i], nums[left]);
		}
    }
};

//Why we still need the redundant 'vec'.. this one is better than mine
class Solution {
public:
    vector<vector<int> > permute(vector<int> &num) {
	    vector<vector<int> > result;
	    
	    permuteRecursive(num, 0, result);
	    return result;
    }
    
    // permute num[begin..end]
    // invariant: num[0..begin-1] have been fixed/permuted
	void permuteRecursive(vector<int> &num, int begin, vector<vector<int> > &result)	{
		if (begin >= num.size()) {
		    // one permutation instance
		    result.push_back(num);
		    return;
		}
		
		for (int i = begin; i < num.size(); i++) {
		    swap(num[begin], num[i]);
		    permuteRecursive(num, begin + 1, result);
		    // reset
		    swap(num[begin], num[i]);
		}
    }
};

File: 046_Permutations.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        def permuteHelper(nums, left, right):
            if left == right:
                res.append(vec[:])
            i = left
            while i < right:
                nums[left], nums[i] = nums[i], nums[left]
                vec[left] = nums[left]
                permuteHelper(nums, left + 1, right)
                nums[i], nums[left] = nums[left], nums[i]
                i += 1
            
        res = list()
        vec = [0] * len(nums)
        permuteHelper(nums, 0, len(nums))
        return res

#What a genious	
def permute(self, nums):
    perms = \[\[\]\]   
    for n in nums:
        new_perms = []
        for perm in perms:
            for i in xrange(len(perm)+1):   
                new_perms.append(perm[:i] + [n] + perm[i:])   ###insert n
        perms = new_perms
    return perms